frameworks/php/cw-2.php

```diff
@@ -0,0 +1,167 @@
+      $this->expect($actual === $expected, "$msg - Expected: " . $this->display($expected) . ", but instead got: " . $this->display($actual));
+    }
+    public function assert_not_equals($actual, $expected, $msg = "Incorrect value returned") {
+      $this->expect($actual !== $expected, "$msg - Algorithm should not have returned: " . $this->display($expected));
+    }
+    public function assert_similar($actual, $expected, $msg = "Actual value did not match expected") {
+      $this->expect($this->check_similar($actual, $expected), "$msg - Expected: " . $this->display($expected) . ", but instead got: " . $this->display($actual));
+    }
+    public function assert_not_similar($actual, $expected, $msg = "Incorrect value returned") {
+      $this->expect(!$this->check_similar($actual, $expected), "$msg - Algorithm should not have returned: " . $this->display($expected));
+    }
+
+    /* Error-Handling Methods */
+    public function expect_error($msg, $fn) {
+      /* Default Message */
+      if (!$msg) $msg = "Expected error was not thrown";
+      $error_thrown = false;
+      try {
+        $fn();
+      } catch (Exception $e) {
+        $error_thrown = true;
+      } finally {
+        $this->expect($error_thrown, $msg);
+      }
+    }
+    public function expect_no_error($msg, $fn) {
+      /* Default Message */
+      if (!$msg) $msg = "Unexpected error was thrown";
+      $error_not_thrown = true;
+      try {
+        $fn();
+      } catch (Exception $e) {
+        $error_not_thrown = false;
+      } finally {
+        $this->expect($error_not_thrown, $msg);
+      }
+    }
+
+    /* Random Output Methods */
+    public function random_number() {
+      return rand(0, 100);
+    }
+    public function random_token() {
+      $length = rand(8, 10);
+      $token = "";
+      for ($i = 0; $i < $length; $i++) {
+        $token .= str_split(Test::token_chars)[floor(lcg_value() * strlen(Test::token_chars))];
+      }
+      return $token;
+    }
+
+    /* Helper Methods (not accessible externally) */
+    protected function check_similar($actual, $expected) {
+      if (!is_array($expected) || !is_array($actual)) return $actual === $expected;
+      foreach ($expected as $key => $value) {
+        if (!$this->check_similar($actual[$key], $expected[$key])) return false;
+      }
+      foreach ($actual as $key => $value) {
+        if (!$this->check_similar($actual[$key], $expected[$key])) return false;
+      }
+      return true;
+    }
+    protected function display($value) {
+      if ($value === true) return "true";
+      if ($value === false) return "false";
+      if ($value === NULL) return "NULL";
+      if (is_string($value)) return '"' . htmlspecialchars($value) . '"';
+      if (!is_array($value)) return $value;
+      if (count($value) === 0) return "array()";
+      $result = "array(";
+      $result .= "<div style='margin-left:20px'>";
```

Comment:
That is true but only when necessary.  As you will have already seen by reading my code, for key operations such as "describe" and "it" blocks, along with passes, fails and errors, only the Codewars output format is outputted.  The only cases where raw HTML is displayed is when my testing framework tries to display actual values as they look in PHP code (e.g. PHP arrays).  As stated in your `README.md` the main purpose of using the Codewars output format is as below:

>The format choosen was originally done so that at any point in time the program could exit while still having readable data.

For the current testing framework I deliberately chose to output HTML *only at places where the code is guaranteed not to unexpectedly exit* so I do not think it is possible that the program exits while outputting HTML leaving unclosed HTML tags.

Your documentation also further states that:

>With the current format there is nothing stopping you from outputing HTML, JS canvas code, etc in order to create a rich and even interactive test result output.

Hopefully my improved version of the PHP testing framework does not output too much HTML :smile:

Cheers,
donaldsebleung